### DISCUSSION TITLE
How to migrate global presets & bookmarks and global dashboards in a different instance?

### DISCUSSION THREAD

**Discussion by daawaan4x** [directus/directus#13056] [Archive]
During development, I create a lot of global presets & bookmarks and global dashboards so our end-users don't have to deal with the hassle anymore of setting up their own ones when viewing data. However, our problem is we can't find a way to programmatically migrate our changes to the production instance in this case. The `directus schema snapshot` command exists but it seems like it only takes a snapshot of your custom models' schema in the database but doesn't include system configurations for directus. Are we missing something? Looking forward for any help!

**Comment by tschortsch** [comment]
Hi @daawaan. We're also working on a directus project and asked ourselves the exact same thing. Till now I think it's only possible to receive the global presets by using the API and export them manually. We're in the middle of implementing a little script which should do that. It's far from ready to be released (and not even in a repository right now) but here is the current state. Maybe you can reuse parts of it:

```typescript
import { IDirectus, TypeMap } from '@directus/sdk';
import { writeFileSync } from 'fs';
import chalk, { Chalk } from 'chalk';

export enum Level {
  INFO = 'INFO',
  SUCCESS = 'SUCCESS',
  WARN = 'WARN',
  ERROR = 'ERROR',
}

const levelToChalkColorMap: {
  [key in Level]: Chalk;
} = {
  [Level.SUCCESS]: chalk.green,
  [Level.INFO]: chalk.blue,
  [Level.WARN]: chalk.yellow,
  [Level.ERROR]: chalk.red,
};
const log = (message: string, level: Level) => {
  console.log(levelToChalkColorMap[level](message));
};

export const exportDefaultPresets = async (
  directus: IDirectus<TypeMap>,
  collection: string,
  target?: string
) => {
  log(`Exporting default presets for collection ${collection}`, Level.INFO);
  try {
    const presets = await directus.presets.readByQuery({
      filter: {
        collection: {
          _eq: collection,
        },
        user: {
          _null: true,
        },
        bookmark: {
          _null: true,
        },
      },
      limit: -1,
    });
    if (!presets.data || presets.data.length === 0) {
      log(`No presets found for ${collection}`, Level.WARN);
      return;
    } else {
      log(
        `Found ${presets.data?.length} ${
          presets.data && presets.data?.length > 1 ? 'presets' : 'preset'
        }`,
        Level.INFO
      );
      const targetFile = `${
        target ? `${target}/` : ''
      }presets-${collection}.json`;
      try {
        writeFileSync(targetFile, JSON.stringify(presets.data, null, 2));
        log(`Sucessfully exported presets to ${targetFile}`, Level.SUCCESS);
      } catch (error) {
        log(`Error while writing ${targetFile}: ${error}`, Level.ERROR);
      }
    }
  } catch (error) {
    log(`Failed to export presets: ${error}`, Level.ERROR);
  }
};
```

Use the helper function like this:

```typescript
import { Directus } from '@directus/sdk';
import { exportDefaultPresets } from './export-default-presets';

const run = async () => {
  const directus = new Directus('http://localhost:8055');
  await directus.auth.login({
    email: 'email',
    password: 'xyz',
  });

  exportDefaultPresets(directus, '<collection_name>', './presets');
};

run();
```
It basically exports all global presets for a given collection and saves them in a file called `presets-<collection_name>.json`.

Maybe the filter query needs to be adjusted to fit more use cases. And we didn't have time to create an import script yet. Hope this helps!

**Comment by rijkvanzanten** [comment]
The above code example relies on the SDK, which under the hood uses the `/presets` endpoints. You should be able to move the presets by reading them from `GET /presets`, and creating the read items with `POST /presets` in the other instance 👍🏻

**Comment by tschortsch** [comment]
We finally found the time to open-source our first utility functions: https://github.com/liip/directus-utilities. We will continue to add helpful utilities to it. Right now we created a separate repository for this to keep up the pace implementing our projects with it but as soon as we have some spare time we'll try to bring it back to the core sdk.

**Reply by LABCAT** [reply]
@tschortsch your repo looks great, any chance you can release an npm package for it?

**Reply by daawaan4x** [reply]
@LABCAT ideally, you should keep custom js scripts instead to tailor it accordingly in your project's needs and use https://github.com/liip/directus-utilities as a basis

**Reply by tschortsch** [reply]
@LABCAT yes this is still on our todo list 😓 But before that we need to cleanup some things and complete the documentation of the available functions. Until then you can install it directly from GitHub via `npm install https://github.com/liip/directus-utilities.git`.

**Reply by LABCAT** [reply]
@tschortsch yes, I ended up installing it in the way you have suggested. The package has been very helpful so far, the presets features are working perfectly. I have submitted a PR that will improve the getPermissionsByRoleIdQuery  function but looks like still a bit of work required to get the import and export of permissions by role working perfectly.

Thank you for making your code open source, has been a great help!

**Reply by tschortsch** [reply]
@LABCAT I finally found the time to release the package to npm: https://www.npmjs.com/package/directus-utilities

**Comment by EdouardDem** [comment]
@daawaan4U 
Hello, I've created a command-line interface (CLI) specifically for this task, available at: https://github.com/tractr/directus-sync. This CLI is designed to export Directus configurations such as schema, `dashboards`, `flows`, `folders`, `operations`, `panels`, `permissions`, `presets`, `roles`, `settings`, `translations`, and `webhooks` into JSON format for storage in your repository. You can then use it to import these configurations back into another Directus instance.
This process is safe and only makes incremental changes. For more information and guidance, I've put together a comprehensive tutorial at: https://medium.com/tractr/streamlining-directus-workflows-with-directus-sync-a-comprehensive-guide-baf33b9cc399.
I hope you find it useful.

**Comment by cronin4392** [comment]
Hey @EdouardDem, I love the directus-sync plugin, it's been a life saver. I can't figure out how to have it sync bookmarks though. My understanding is it's only meant to sync schemas, not the data itself, so my actual bookmarks don't get synced. Am I missing something?
